# ...

# Ensemble ML with Wine Dataset

## Imports

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.ensemble import (
    RandomForestClassifier,
)
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import (
    confusion_matrix,
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
)

## Introduction

In this project, I’m exploring how well different machine learning models can predict the quality of red wine based on its chemical properties. The dataset includes things like acidity, sulphates, alcohol percentage, and more, along with a quality score rated by human tasters. My main goal in this notebook is to compare different models, understand why some work better than others, and get a clearer idea of what makes certain models a good fit for this kind of data.

## Section 1. Load and Inspect the Data

In [23]:
df = pd.read_csv("wine+quality/winequality-red.csv",sep=";")

In [24]:
# Display structure and first few rows
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5


## Project 2. Prepare the Data

In [25]:
def quality_to_label(q):
    if q <= 4:
        return "low"
    elif q <= 6:
        return "medium"
    else:
        return "high"

In [26]:
df["quality_label"] = df["quality"].apply(quality_to_label)

In [27]:
def quality_to_number(q):
    if q <= 4:
        return 0
    elif q <= 6:
        return 1
    else:
        return 2

In [28]:
df["quality_numeric"] = df["quality"].apply(quality_to_number)

## Section 3. Feature Selection and Justification

In [29]:
X = df.drop(columns=["quality", "quality_label", "quality_numeric"])
y = df["quality_numeric"]

## Section 4. Split the Data into Train and Test

In [30]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size = 0.2, random_state = 42, stratify = y
)

## Section 5. Evaluate Model Performance (Choose 2)

In [31]:
def evaluate_model(name, model, X_train, y_train, X_test, y_test, results):
    model.fit(X_train, y_train)

    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)

    train_acc = accuracy_score(y_train, y_train_pred)
    test_acc = accuracy_score(y_test, y_test_pred)
    train_f1 = f1_score(y_train, y_train_pred, average="weighted")
    test_f1 = f1_score(y_test, y_test_pred, average="weighted")

    print(f"\n{name} Results")
    print("Confusion Matrix (Test):")
    print(confusion_matrix(y_test, y_test_pred))

    # Fix: SyntaxError: invalid decimal literal
    print(f"Train Accuracy: {train_acc}, Test Accuracy: {test_acc}")
    print(f"Train F1 Score: {train_f1}, Test F1 score: {test_f1}")

    results.append(
        {
            "Model": name,
            "Train Accuracy": train_acc,
            "Test Accuracy": test_acc,
            "Train F1": train_f1,
            "Test F1": test_f1
        }
    )


In [32]:
results = []

# 1. Random Forest
evaluate_model(
    "Random Forest (100)",
    RandomForestClassifier(n_estimators=100, random_state=42),
    X_train,
    y_train,
    X_test,
    y_test,
    results,
)

# 2. Random Forest (200, max depth=10) 
evaluate_model(
    "Random Forest (200, max_depth=10)",
    RandomForestClassifier(n_estimators=200, max_depth=10, random_state=42),
    X_train,
    y_train,
    X_test,
    y_test,
    results,
)


Random Forest (100) Results
Confusion Matrix (Test):
[[  0  13   0]
 [  0 256   8]
 [  0  15  28]]
Train Accuracy: 1.0, Test Accuracy: 0.8875
Train F1 Score: 1.0, Test F1 score: 0.8660560842649911

Random Forest (200, max_depth=10) Results
Confusion Matrix (Test):
[[  0  13   0]
 [  0 255   9]
 [  0  16  27]]
Train Accuracy: 0.9757623143080532, Test Accuracy: 0.88125
Train F1 Score: 0.9744822364939886, Test F1 score: 0.8596432366257046


## Section 6: Compare Results

In [33]:
# Create a table of results 
results_df = pd.DataFrame(results)

print("\nSummary of All Models:")
display(results_df)


Summary of All Models:


,Model,Train Accuracy,Test Accuracy,Train F1,Test F1
0,Random Forest (100),1.000000,0.88750,1.000000,0.866056
1,"Random Forest (200, max_depth=10)",0.975762,0.88125,0.974482,0.859643


## Section 7. Conclusions and Insights

Across the models I tested, tree-based ensemble methods performed the best at predicting red wine quality. The 100-tree Random Forest achieved the strongest results, with 0.8875 test accuracy and a 0.866 weighted F1-score, slightly outperforming the 200-tree, max-depth-10 version. Increasing complexity beyond this did not improve performance.

A major reason Random Forest works so well on this dataset is that wine quality depends on non-linear relationships and feature interactions: for example, alcohol interacts with acidity, sulphates, and volatile acidity in ways that aren’t well captured by a linear decision boundary. Tree models naturally learn thresholds and interaction rules (“if alcohol > X *and* volatile acidity < Y → higher quality”), making them a better match for the underlying chemistry. Random Forest improves on a single tree by averaging many trees trained on different subsets of features and data, which reduces variance and stabilizes performance. This behavior matches what classmates found as well: tree ensembles consistently outperform logistic regression, KNN, and single decision trees.

Given the results, the Random Forest with 100 trees is the best model in this notebook: accurate, robust, and interpretable through feature importance. If I were pushing for competition-level performance, the next steps would include trying Gradient Boosting / XGBoost, more extensive hyperparameter tuning, addressing class imbalance, and possibly stacking multiple ensembles.

In summary, ensemble tree methods succeed because they capture the complex, non-linear structure of the wine data, and both my results and others’ support this conclusion.